add-on/src/lib/redirect-handler/blockOrObserve.ts

```diff
@@ -119,6 +134,84 @@ function constructRegexFilter ({ originUrl, redirectUrl }: redirectHandlerInput)
+    while (subdomain.length > 0) {
+      const subdomainPart = subdomain.shift()
+      const commonStaticUrlStart = `^${originURL.protocol}\\:\\/\\/`
+      const commonStaticUrlEnd = `\\.${escapeURLRegex(staticUrl.join('.'))}\\/${RULE_REGEX_ENDING}`
+      if (isIPFS.cid(subdomainPart as string)) {
+        // We didn't find a namespace, but we found a CID
+        // e.g. https://bafybeib3bzis4mejzsnzsb65od3rnv5ffit7vsllratddjkgfgq4wiamqu.on.fleek.co
+        regexFilter = `${commonStaticUrlStart}(.*?)${commonStaticUrlEnd}`
+        regexSubstitution = redirectUrl
+          .replace(subdomainPart as string, '\\1') // replace CID
+          .replace(new RegExp(`${originURL.pathname}?$`), '\\2') // replace path
+
+        break
+      }
+      if (DEFAULT_NAMESPACES.has(subdomainPart as string)) {
+        // We found a namespace, this is going to match group 2, i.e. namespace.
+        // e.g https://bafybeib3bzis4mejzsnzsb65od3rnv5ffit7vsllratddjkgfgq4wiamqu.ipfs.dweb.link
+        regexFilter = `${commonStaticUrlStart}(.*?)\\.(${[...DEFAULT_NAMESPACES].join('|')})${commonStaticUrlEnd}`
+
+        regexSubstitution = redirectUrl
+          .replace(subdomain.reverse().join('.'), '\\1') // replace subdomain or CID.
+          .replace(`/${subdomainPart as string}/`, '/\\2/') // replace namespace dynamically.
+
+        const pathWithSearch = originURL.pathname + originURL.search
+        if (pathWithSearch !== '/') {
+          regexSubstitution = regexSubstitution.replace(pathWithSearch, '/\\3') // replace path
+        } else {
+          regexSubstitution += '\\3'
+        }
+
+        break
+      }
+      // till we find a namespace or CID, we keep adding subdomains to the staticUrl.
+      staticUrl.unshift(subdomainPart as string)
+    }
+
+    if (regexFilter !== origRegexFilter) {
+      // we found a valid regexFilter, so we can return.
+      return { regexSubstitution, regexFilter }
+    } else {
+      // we didn't find a valid regexFilter, so we can return the default.
+      regexFilter = originUrl
+    }
+  }
+
+  // if the namespaces are the same, we can generate simpler regex.
+  // The only value that needs special handling is the `uri` param.
+  // TODO: Remove this check, as `uri` param is deprecated.
+  if (
+    DEFAULT_NAMESPACES.has(originNS) &&
+    DEFAULT_NAMESPACES.has(redirectNS) &&
+    originNS === redirectNS &&
+    originURL.searchParams.get('uri') == null
+  ) {
+    // A redirect like
+    // https://ipfs.io/ipfs/QmZMxU -> http://localhost:8080/ipfs/QmZMxU
+    const [originFirst, originLast] = originUrl.split(`/${originNS}/`)
+    const defaultNSRegexStr = `(${[...DEFAULT_NAMESPACES].join('|')})`
+    regexFilter = `^${escapeURLRegex(originFirst)}\\/${defaultNSRegexStr}\\/${RULE_REGEX_ENDING}`
+      .replace(/https?/ig, 'https?')
+    regexSubstitution = redirectUrl
+      .replace(`/${redirectNS}/`, '/\\1/')
+      .replace(originLast, '\\2')
+    return { regexSubstitution, regexFilter }
```

Comment:
It would be great to encapsulate the different return values throughout this function so we could name them and break up this function a bit more.